KEP-5051: Add Server Side Apply: Unsetting Fields KEP (#5052)
kubernetes/enhancements: keps/sig-api-machinery/5051-server-side-apply-unsetting-fields/README.md

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes an improvement to Server Side Apply to allow apply configurations to declare the
intent that a field should not be set.

## Motivation

As of Kubernetes 1.32, Server side apply lacks a direct way to unset fields.

If a field manager exclusively owns a field, the field manage may unset the field by applying a
configuring that omits the field. But if the field manager does not exclusively own the field, there
is no way for the field manager to unset the field using an apply request.

Additionally, there is no way for a field manager to own a field that is unset.  This prevents a
field manager from expressing the opinion that a field should be unset, and to force a conflict if
another field manager attempts to set the field.

### Goals

- Introduce a way to express that a field should be unset in a Server Side Apply configuration.
- Introduce field ownership of unset fields into the managedFields of Server Side Apply.

### Non-Goals

> N/A

Reviewer: This _is_ applicable, but there aren't any.
```suggestion
_none_
```

## Proposal

Introduce a "marker" value that may be used in apply configurations to indicate that a value should
be unset.

> The proposed marker value is: `{k8s_io__value: unset}`

Reviewer 2: Are there any other k8s-features/SSA-features that use a marker value like {k8s_io__value: unset}?  Might be of use to show that there is some precedent for the format and the approach if so.

Author: There is not. We're diverging from the "all apply configurations are valid subsets of the full resource representation" here.

Reviewer: Something to call out very thoroughly in docs, then.
- People have previously not need to know if their YAML file represents an incomplete object you can submit using a **create** verb or an apply configuration for SSA.
- We'll need to define which of these two meanings aligns with our specialist jargon term _manifest_

Reviewer: Maybe we could accept `{k8s_io__value: unset}` for **create** / `PUT` too?

Reviewer: Related, I think: if you have a manifest rather than an apply configuration, what `kubectl` subcommand do you use to ensure that the manifest represents your declared, desired state?

(I think I know the answer and I don't like it).

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Apply configuration:

```yaml
spec:
  field: "xyz"
```

Apply conflicts:

```yaml
field owned by fieldManager1
```

#### Apply (force=false) to unset a field that is owned but already unset

Existing value:

```yaml
spec:
```

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Apply configuration:

```yaml
spec:
  field: {k8s_io__value: unset}
```

Result value:

```yaml
spec:
```

Result field management:

```yaml
Shared ownership.

fieldManager1: 
  spec.field

fieldManager2: 
  spec.field
```

#### Apply to a field with a default value

Existing value:

```yaml
spec:
  field: defaultValue
```

Existing field management:

```yaml
fieldManager1:
  # no owned fields
```

Force apply configuration:

```yaml
spec:
  field: {k8s_io__value: unset}
```

Result value:

```yaml
spec:
  field: defaultValue
```

Result field management:

```yaml
fieldManager2:
  spec.field
```

### Notes/Constraints/Caveats (Optional)

#### Defaulting considerations

1. A field that is not explicitly set, but is defaulted, is unowned (existing behavior)
1. If a field is unset using a marker, defaulting still applies

>   - Consequence: A field manager that owns a unset value ends up owning the defaulted value
>   - Consequence: If two field managers unset the same value, they share ownership of the field
>   - Caveat: Non-declarative defaulting, such as defaulting that is performed by the strategy, or
>     admission control, is not detectable by server side apply, and will result in conflicts between
> 	field managers even for cases where the defaulting SHOULD result in shared field ownership.
> 	This is a pre-existing problem between defaulting and server side apply but is more likely
> 	to occur with this enhancement.

Reviewer 3: ```suggestion
   - Consequence: A field manager that owns a unset value ends up owning the defaulted value
   - Consequence: If two field managers unset the same value, they share ownership of the field
   - Caveat: Non-declarative defaulting, such as defaulting that is performed by the strategy, or
    admission control, is not detectable by server side apply, and will result in conflicts between
	field managers even for cases where the defaulting SHOULD result in shared field ownership.
	This is a pre-existing problem between defaulting and server side apply but is more likely
	to occur with this enhancement.
```
nit: one more indent to continue the order number

> 1. listType=map key fields that are defaulted MUST be respected extracting unset value markers from

Reviewer 3: ```suggestion
1. listType=map key fields that are defaulted MUST be respected when extracting unset value markers from
```
Do you mean this?

### Unset marker escaping

We will NOT support escaping of the marker.

We considered this and started to prototype an implementation, but:

- Adding a performant unescaping pass to structured-merge-diff is complex. It effectively doubles
  the implementation effort of this KEP.
- It's hard to imagine where escaping this key would actually be needed. If we really need the ability
  to "apply to an apply configuration" in the future, look into options, but building this without
  a plausible use case does not seem necessary.

### High level implementation plan

In kubernetes-sigs/structured-merge-diff:

- An option will be added to apply configuration validation to allow unset field markers to pass
  validation.
- A new `TypedValue.ExtractMarkers()` function will be added that separates the markers from an
  apply configuration and returns them as a field set.
- An option will be added to `Updater.Apply` to allow unset fields.

In kubernetes/kubernetes:

- A feature gate will be added and used to enable support for unset field markers on
  `Updater.Apply`.
- MutatingAdmissionPolicy (Alpha feature) will be modified to always allow unset field markers.

>   The use of unset field markers will NOT be feature gated in this alpha feature.

Reviewer 3: Though it is stated above that MutatingAdmissionPolicy won't graduate to Beta before UnsettingField, would it still introduce some risk that falsely allowing unset field markers when UnsettingField feature is disabled?

And, OOC, is it normal to handle relation between features like this(have a graduate dependency)?

Author: > Though it is stated above that MutatingAdmissionPolicy won't graduate to Beta before UnsettingField, would it still introduce some risk that falsely allowing unset field markers when UnsettingField feature is disabled?

We intend to unconditionally allow MutatingAdmissionPolicy to allow fields to be unset when the MutatingAdmissionPolicy feature gate is enabled. The graduation restriction is purely to ensure that the unsetting fields is not being by a feature that claims a stability level above the stability that we claim unsetting fields is at.

While we could configure MutatingAdmissionPolicy to not support unsetting fields unless both the MutatingAdmissionPolicy and UnsettingField feature gates are enabled, we don't really need to. What we're effectively doing is saying "MutatingAdmissionPolicy has unsetting field support. We making it an intrinsic part of the feature in alpha.  Can we guard against any risk this might cause by restricting MutatingAdmissionPolicy promotion to not progress beyond UnsettingField"

> And, OOC, is it normal to handle relation between features like this(have a graduate dependency)?

Yes, we've go graduation dependencies between other KEPs.  It's not uncommon.

    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.

> -->

Reviewer: Are there any?

> ## Infrastructure Needed (Optional)
> 
> <!--
> Use this section if you need things from the project/SIG. Examples include a
> new subproject, repos requested, or GitHub details. Listing these here allows a
> SIG to get the process for these resources started right away.
> -->

Reviewer: ```suggestion
```

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes an improvement to Server Side Apply to allow apply configurations to declare the
intent that a field should not be set.

## Motivation

As of Kubernetes 1.32, Server side apply lacks a direct way to unset fields.

> If a field manager exclusively owns a field, the field manage may unset the field by applying a

Reviewer 4: ```suggestion
If a field manager exclusively owns a field, the field manager may unset the field by applying a
```

> configuring that omits the field. But if the field manager does not exclusively own the field, there

Reviewer 4: ```suggestion
configuration that omits the field. But if the field manager does not exclusively own the field, there
```

  spec.field
```

#### Apply to a field with a default value

Existing value:

```yaml
spec:
  field: defaultValue
```

Existing field management:

```yaml
fieldManager1:
  # no owned fields
```

Force apply configuration:

```yaml
spec:
  field: {k8s_io__value: unset}
```

Result value:

```yaml
spec:
  field: defaultValue
```

Result field management:

```yaml
fieldManager2:
  spec.field
```

### Notes/Constraints/Caveats (Optional)

#### Defaulting considerations

1. A field that is not explicitly set, but is defaulted, is unowned (existing behavior)
1. If a field is unset using a marker, defaulting still applies
   - Consequence: A field manager that owns a unset value ends up owning the defaulted value
   - Consequence: If two field managers unset the same value, they share ownership of the field
   - Caveat: Non-declarative defaulting, such as defaulting that is performed by the strategy, or
    admission control, is not detectable by server side apply, and will result in conflicts between
	field managers even for cases where the defaulting SHOULD result in shared field ownership.
	This is a pre-existing problem between defaulting and server side apply but is more likely

> 	to occur with this enhancement.

Reviewer 4: Can you give a more concrete example? When you say "not detectable", what does that mean? Which kind of conflict? An update conflict deep in the storage stack?

Reviewer 4: In general, I understand that admission defaulting behaves different than deserializer defaulting. What does that mean for your usecases above?

Author: Good question. The case I'm thinking of is:

- FieldManager 1: Applies field X to A
  - Apply succeeds, FieldManager 1 owns X
- Field X is set to B _after_ deserialization defaulting (typically the strategy or admission chain)
- FieldManager 2: Applies field X to A
- Apply conflict

This differs from deserialization defaulting, which will result in shared ownership of the field.

I'll find or open an issue explaining and link to it from this section of the KEP.

Author: Actually, I think this can only happen today with mutating admission, but could happen in the future as a result of defaulting as a result of this feature.  https://github.com/kubernetes/kubernetes/issues/129960 has details. I'l include a link in the KEP now.

## Summary

This KEP proposes an improvement to Server Side Apply to allow apply configurations to declare the
intent that a field should not be set.

## Motivation

As of Kubernetes 1.32, Server side apply lacks a direct way to unset fields.

If a field manager exclusively owns a field, the field manage may unset the field by applying a
configuring that omits the field. But if the field manager does not exclusively own the field, there
is no way for the field manager to unset the field using an apply request.

Additionally, there is no way for a field manager to own a field that is unset.  This prevents a
field manager from expressing the opinion that a field should be unset, and to force a conflict if
another field manager attempts to set the field.

### Goals

- Introduce a way to express that a field should be unset in a Server Side Apply configuration.
- Introduce field ownership of unset fields into the managedFields of Server Side Apply.

### Non-Goals

_none_

## Proposal

Introduce a "marker" value that may be used in apply configurations to indicate that a value should
be unset.

The proposed marker value is: `{k8s_io__value: unset}`

> For example, given a resource with a field owned by field manager "mgr1":

Reviewer: ```suggestion
For example, given an object with a field owned by field manager "mgr1":
```

```yaml
apiVersion: stable.example.com/v1
kind: Example
metadata:
  name: example1
  managedFields:
  - manager: mgr1
    operation: Apply
    …
    fieldsV1:
      f:spec:
        f:field {}
spec:
  field: "xyz"
```

Field manager "mgr2" may unset the field by force applying a apply configuration like:

```yaml
apiVersion: stable.example.com/v1
kind: Example
metadata:
  name: example1
spec:
  field: {k8s_io__value: unset}
```

After the configuration is applied field will be unset and will be owned by "mgr2".

```yaml
apiVersion: stable.example.com/v1
kind: Example
metadata:
  name: example1
  managedFields:
  - manager: mgr2
    operation: Apply
    …
    fieldsV1:
      f:spec:
        f:field {} # mgr2 owns this field even though it is unset
spec:
  # empty
```

> ### User Stories (Optional)

Reviewer: ```suggestion
### User Stories
```
We have some

Mitigations: 

- Extensive testing. See Test Plan section.
- The new functionality will be isolated and gated to ensure the pre-existing behavior isolated
  preserved when the feature is off.

### Risk: The additional processing of apply requests negatively impacts performance

Mitigations:

- This change will be benchmarked.
- Apply requests that do not unset fields will only require the read-only check to search for unset
  field markers. All other new code will be skipped if no unset field markers are found.
- This change will leverage the optimizations used in structured-merge-diff to minimize allocations
  when traversing managedFields.

## Design Details

#### Admission Control

MutatingAdmissionPolicy will support the use of this feature.

For example, an ApplyConfiguration mutation in a MutatingAdmissionPolicy may unset a field:

```cel
Object{
  spec: Object.spec{
    template: Object.spec.template{
      spec: Object.spec.template.spec{
        volumes: [Object.spec.template.spec.volumes{
          name: "y",
          k8s_io__value: "unset",
        }]
      }
    }
}
```

This mutation unsets a field from the request.

> This mutation DOES NOT result in the managedFields tracking ownership of the unset field.

Reviewer: (nit)
```suggestion
This mutation **DOES NOT** result in the `managedFields` tracking ownership of the unset field.
```

> MutatingAdmissionPolicy is an Alpha feature and will support unsetting fields unconditionally (even

Reviewer: ```suggestion
At the time of KEP creation, MutatingAdmissionPolicy was an Alpha API; you can use a MutatingAdmissionPolicy to unset fields unconditionally (even
```

if the ServerSideApplyUnsettingFields feature gate is off). MutatingAdmissionPolicy will not

> graduate to Beta before ServerSideApplyUnsettingFields to ensure we do not limit our ability

Reviewer: ```suggestion
graduate to Beta before the this KEP and the `ServerSideApplyUnsettingFields` feature gate; we are phasing things
that way to ensure we do not limit our ability
```

to respond to community feedback for this enhancement.

### Type-safe apply configuration bindings

Before graduation to beta, we will add unsetting field support to `applyconfiguration-gen`.

Our goal is to add `Unset<FieldName>()` functions to the generated types. This may require
the introduction of custom marshalling to the apply configuration types.

### Unset marker escaping

> We will NOT support escaping of the marker.

Reviewer: ```suggestion
We will **NOT** support escaping of the marker.
```

We considered this and started to prototype an implementation, but:

- Adding a performant unescaping pass to structured-merge-diff is complex. It effectively doubles
  the implementation effort of this KEP.
- It's hard to imagine where escaping this key would actually be needed. If we really need the ability
  to "apply to an apply configuration" in the future, look into options, but building this without

>   a plausible use case does not seem necessary.

Reviewer: Could list this under Alternatives.

### High level implementation plan

In kubernetes-sigs/structured-merge-diff:

- An option will be added to apply configuration validation to allow unset field markers to pass
  validation.
- A new `TypedValue.ExtractMarkers()` function will be added that separates the markers from an
  apply configuration and returns them as a field set.
- An option will be added to `Updater.Apply` to allow unset fields.

In kubernetes/kubernetes:

- A feature gate will be added and used to enable support for unset field markers on
  `Updater.Apply`.
- MutatingAdmissionPolicy (Alpha feature) will be modified to always allow unset field markers.

>   The use of unset field markers will NOT be feature gated in this alpha feature.

Reviewer: ```suggestion
  The use of unset field markers will **NOT** be feature gated in this alpha feature.
```

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

> <!--
> Pick one of these and delete the rest.
> 
> Documentation is available on [feature gate lifecycle] and expectations, as
> well as the [existing list] of feature gates.
> 
> [feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
> [existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
> -->

Reviewer: ```suggestion
```

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ServerSideApplyUnsettingFields

>   - Components depending on the feature gate: apiservers

Reviewer: ```suggestion
  - Components depending on the feature gate: kube-apiserver
```

```yaml
apiVersion: stable.example.com/v1
kind: Example
metadata:
  name: example1
  managedFields:
  - manager: mgr2
    operation: Apply
    …
    fieldsV1:
      f:spec:
        f:field {} # mgr2 owns this field even though it is unset
spec:
  # empty
```

### User Stories

#### Apply (force=true) to unset an owned listType=map element

Existing value:

```yaml
field: [
  {name: "a", value: 1}, 
  {name: "b", value: 2}
]
```

Existing field management:

```yaml
fieldManager1: 
  spec.field[name=a]
  spec.field[name=b]
```

Force apply configuration:

```yaml
field: [
  {name: "b", k8s_io__value: unset}
]
```

Result value:

```yaml
field: [
  {name: "a", value: 1}
]
```

Result field management:

```yaml
fieldManager1: 
  spec.field[name=a]

fieldManager2: 
  spec.field[name=b]
```

#### Apply (force=true) to unset an owned granular map value

Existing value:

```yaml
field: {"a": 1, "b": 2}
```

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Force apply configuration:

```yaml
field: {"b": {k8s_io__value: unset}}
```

Result value:

```yaml
field: {"a": 1}
```

Result field management:

```yaml
fieldManager2:

>   field

Author: Both granular and atomic maps are supported.  This shows an example of an atomic map.  I can clarify and also include an example of a granular map.

Implementation test suite covers both.

```

#### Apply (force=false) to an owned field

Existing value:

```yaml
spec:
  field: 1
```

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Apply configuration:

```yaml
spec:
  field: {k8s_io__value: unset}
```

Apply conflicts:

```yaml
field owned by fieldManager1
```

#### Apply (force=false) to a unset field that is owned

Existing value:

```yaml
spec:
  # empty
```

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Apply configuration:

```yaml
spec:
  field: "xyz"
```

Apply conflicts:

```yaml
field owned by fieldManager1
```

#### Apply (force=false) to unset a field that is owned but already unset

Existing value:

```yaml
spec:
```

Existing field management:

```yaml
fieldManager1: 
  spec.field
```

Apply configuration:

```yaml
spec:

>   field: {k8s_io__value: unset}

Reviewer 5: when encoded in JSON or CBOR, this will be a string containing this?  Actually what format is this?  it's not quite JSON.

Reviewer: This looks like YAML to me. I think the JSON would be something like:
```yaml
  "spec": {
    "field": {
      "k8s_io__value": "unset"
    },
```

Author: That's right.  I could switch this example to be more idiomatic YAML if that helps:

```yaml
spec:
  field:
    k8s_io__value: unset
```

Reviewer 5: > This looks like YAML to me. I think the JSON would be something like:

Oh, my eye didn't read it that way.  That means that it is logically impossible to escape the value for a  piece of say integer configuration exposed from one resource to set a value in another.  For instance, a controller of deployments would be unable to express "must be empty" in its API.

That seems really problematic.

  spec.field
```

#### Apply to a field with a default value

Existing value:

```yaml
spec:
  field: defaultValue
```

Existing field management:

```yaml
fieldManager1:
  # no owned fields
```

Force apply configuration:

```yaml
spec:
  field: {k8s_io__value: unset}
```

Result value:

```yaml
spec:
  field: defaultValue
```

Result field management:

```yaml
fieldManager2:
  spec.field
```

### Notes/Constraints/Caveats (Optional)

#### Defaulting considerations

1. A field that is not explicitly set, but is defaulted, is unowned (existing behavior)
1. If a field is unset using a marker, defaulting still applies
   - Consequence: A field manager that owns a unset value ends up owning the defaulted value
   - Consequence: If two field managers unset the same value, they share ownership of the field
   - Caveat: Non-declarative defaulting, such as defaulting that is performed by the strategy, or
    admission control, is not detectable by server side apply, and will result in conflicts between
	field managers even for cases where the defaulting SHOULD result in shared field ownership.
	This is a pre-existing problem between mutating admission and server side apply but

> 	will also become possible via defaulting of unset fields with this enhancement.

Author: Sure.  I'll port in some of the data from the linked issue.

if the ServerSideApplyUnsettingFields feature gate is off). MutatingAdmissionPolicy will not
graduate to Beta before the this KEP and the `ServerSideApplyUnsettingFields` feature gate; we are phasing things
that way to ensure we do not limit our ability
to respond to community feedback for this enhancement.

### Type-safe apply configuration bindings

Before graduation to beta, we will add unsetting field support to `applyconfiguration-gen`.

Our goal is to add `Unset<FieldName>()` functions to the generated types. This may require
the introduction of custom marshalling to the apply configuration types.

### Unset marker escaping

<<<<<<< Updated upstream
We will **NOT** support escaping of the marker.

We considered this and started to prototype an implementation, but:

- Adding a performant unescaping pass to structured-merge-diff is complex. It effectively doubles
  the implementation effort of this KEP.
- It's hard to imagine where escaping this key would actually be needed. If we really need the ability
  to "apply to an apply configuration" in the future, look into options, but building this without

>   a plausible use case does not seem necessary.

Author: Unescaping `__k8s_io__value` to `k8s_io__value` might not be that bad in the current implementation.  Okay if I make this a beta graduation criteria?  (It seems valuable for production grade but not essential for proving the idea at the alpha level?)

>   a plausible use case does not seem necessary.

Reviewer 5: > Unescaping `__k8s_io__value` to `k8s_io__value` might not be that bad in the current implementation. Okay if I make this a beta graduation criteria? (It seems valuable for production grade but not essential for proving the idea at the alpha level?)

Must be present before going beta, but yes I think we could try alpha without it.

>   a plausible use case does not seem necessary.

Author: :+1: I've almost got it working, I'll target this for the 1st alpha, possibly as a separate PR.
